File: haskell/008/Main.hs
{-
    ProjectEuler 008
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

import Data.Char

main :: IO ()
main = do
    contents <- readFile "./digits.txt"
    let digits = map digitToInt . head $ lines contents
    print $ calc 13 0 digits (length digits)

calc :: Int -> Int -> [Int] -> Int -> Int
calc n p digits l
    | l < n     = p
    | otherwise = let q = product $ take n digits in calc n (max p q) (tail digits) (l - 1)

File: scala/008/Main.scala
/*
 * ProjectEuler 008
 * Scala / Main.scala
 * copyright (c) 2015 Susisu
 *
 * Scala 2.11.4
 * `scalac Main.scala -d bin`
 */

import scala.io.Source

object Main {
    def main(args: Array[String]) {
        val contents = Source.fromFile("./digits.txt");
        val digits = contents.getLines.toList.head.split("").map(_.toInt);
        val n = 13;
        val len = digits.length;
        var max: Long = 0;
        for (i <- 0 to len - n) {
            var p: Long = 1;
            for (j <- 0 until n) {
                p *= digits(i + j);
            }
            if (p > max) {
                max = p;
            }
        }
        println(max);
    }
}
